### PR TITLE
Fix the NPE for DISTINCT aggregation function when no record is selected

### PR DESCRIPTION
**Pull request by Jackie-Jiang**
When no record is selected for the whole table, column data type will be STRING for all distinct columns (by Pinot convention)

### CODE REVIEW

File: pinot-core/src/test/java/org/apache/pinot/queries/DistinctQueriesTest.java [hunk 1]
```diff
@@ -685,6 +694,18 @@ public void testDistinctWithFilter()
         sortedResults.add(new Record(new Object[]{"California", "San Mateo", 400000}));
         sortedResults.add(new Record(new Object[]{"California", "Sunnyvale", 300000}));
         runQueryInterSegmentWithOrderBy(orderByQuery, sortedResults, new String[]{"State", "City", "SaleAmount"});
+
+        String emptyResultQuery =
```
Comment (siddharthteotia) [inline pinot-core/src/test/java/org/apache/pinot/queries/DistinctQueriesTest.java]: I think this is the scenario when the overall result set of DISTINCT query at table level is empty and was throwing NPE as DistinctTable was null. 

Can we please also add a test where the overall result at table level is non-empty but one or more segment returns empty?

Then we have everything  covered.

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AggregationFunctionFactory.java [hunk 2]
```diff
@@ -112,8 +111,10 @@ public static AggregationFunction getAggregationFunction(AggregationInfo aggrega
           case DISTINCTCOUNTRAWHLLMV:
             return new DistinctCountRawHLLMVAggregationFunction();
           case DISTINCT:
+            Preconditions.checkState(brokerRequest != null,
+                "Broker request must be provided for 'DISTINCT' aggregation function");
             return new DistinctAggregationFunction(AggregationFunctionUtils.getColumn(aggregationInfo),
-                brokerRequest != null ? brokerRequest.getLimit() : SelectAstNode.DEFAULT_RECORD_LIMIT, brokerRequest.getOrderBy());
+                brokerRequest.getOrderBy(), brokerRequest.getLimit());
```
Comment (siddharthteotia) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AggregationFunctionFactory.java]: I am unable to recall why I had this if check. I think there was a code-path where null broker request could have been passed. It is not possible right?

File: pinot-core/src/test/java/org/apache/pinot/queries/DistinctQueriesTest.java (hunk 1 above)
Comment (siddharthteotia) [reply pinot-core/src/test/java/org/apache/pinot/queries/DistinctQueriesTest.java]: Now that I think more about it, I don't think this solution will work for the case when result from all segments on a single server is empty but across the servers is non empty. 

The current solution would have sent. a Distinct table from the server with string types. However when the broker reduces distinct tables across server, it might choose the result column types as string if the server that sent empty distinct tables happens to be the 0th server. The broker has to decide the result column types for.the final response and it does that upfront by looking at the column types from first server's table since it assumes they will be the same (they should be in general cases except for this corner case). This will be wrong.

Comment (siddharthteotia) [reply pinot-core/src/test/java/org/apache/pinot/queries/DistinctQueriesTest.java]: We need to introduce some hacky way of knowing this scenario in the broker.reduce code.

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AggregationFunctionFactory.java (hunk 2 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AggregationFunctionFactory.java]: The existing code never works for `null` brokerRequest because `brokerRequest.getOrderBy()` will directly throw NPE. The broker aggregation reducer might call this without brokerRequest, but distinct is not using that code path to reduce the result.
I added TODOs in `AggregationFunctionUtils` to always pass broker request for all code paths

File: pinot-core/src/test/java/org/apache/pinot/queries/DistinctQueriesTest.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-core/src/test/java/org/apache/pinot/queries/DistinctQueriesTest.java]: As long as it is using the `DistinctAggregationFunction.merge()` to merge the DistinctTables, it is handled. Check the fix in  `DistinctAggregationFunction.merge()`.
Let me try to enhance the test to cover that.

**Comment by Jackie-Jiang** [comment]
Discussed offline. Will modify the test in another pr to cover different scenarios of empty response.
